token-tracker/src/truth-sources/lp-tokens.js

```diff
@@ -0,0 +1,267 @@
+
+    return lpBalanceByStaker
+  }
+
+  /**
+   * Calculates KEEP for all LP KEEP-ETH / KEEP-TBTC stakers.
+   *
+   * @param {Map<Address, BN>} stakersBalances LP KEEP-ETH / KEEP-TBTC Token amounts by stakers
+   *
+   * @return {Map<Address,BN>} KEEP Tokens in LP KEEP-ETH / KEEP-TBTC at the target block
+   */
+  async calcKeepInStakersBalances(stakersBalances) {
+    logger.info(`check token stakers at block ${this.targetBlock}`)
+
+    const keepInLpByStakers = new Map()
+
+    // Retrieve current pair data
+    const pairData = await getPairData(this.liquidityStaking.lpPairAddress)
+    for (const [stakerAddress, lpBalance] of stakersBalances.entries()) {
+      const keepInLPToken = await this.calcKeepTokenfromLPToken(
+        lpBalance,
+        pairData
+      )
+      keepInLpByStakers.set(stakerAddress, keepInLPToken)
+
+      logger.info(
+        `Staker: ${stakerAddress} - LP Balance: ${lpBalance} - KEEP in LP: ${keepInLPToken}`
+      )
+    }
+
+    logger.info(
+      `found ${keepInLpByStakers.size} stakers at block ${this.targetBlock}`
+    )
+
+    dumpDataToFile(
+      keepInLpByStakers,
+      this.liquidityStaking.keepInLpTokenFilePath
+    )
+
+    return keepInLpByStakers
+  }
+
+  /**
+   * Calculates amount of KEEP token which makes a KEEP-ETH / KEEP-TBTC Uniswap pair.
+   * A Uniswap LP pair is a bookeeping tool to keep track of how much the liquidity
+   * stakers are owed. They store two assets of equivalent value of each, ex. KEEP-ETH.
+   * This means that the value of KEEP owned is dependent on the ratio of staked
+   * LP tokens and the total LP supply. Ratio between LP tokens and KEEP tokens
+   * should be equal:
+   * LP_staker_balance / LP_total_supply_pool == KEEP_staker_owed / KEEP_total_liquidity_pool
+   * Now, the number of KEEP tokens which makes a KEEP-ETH pair can be calculated
+   * using the following equation:
+   * KEEP_staker_owed = (LP_staker_balance * KEEP_total_liquidity_pool) / LP_total_supply_pool
+   * where:
+   * LP_staker_balance is retrieved from LPRewardsContract
+   * KEEP_total_liquidity_pool is queried from Uniswap API - pairData.reserve0
+   * LP_total_supply_pool is queried from Uniswap API - pairData.totalSupply
+   *
+   * References:
+   * Uniswap API: https://uniswap.org/docs/v2/API/queries/#pair-data
+   * Returns in Uniswap: https://uniswap.org/docs/v2/advanced-topics/understanding-returns/
+   *
+   * @param {BN} lpBalance LP amount staked by a staker in a LPRewardsContract
+   * @param {PairData} pairData KEEP-ETH / KEEP-TBTC pair data fetched from Uniswap
+   *
+   * @return {BN} KEEP token amounts in LP token balance
+   */
+  async calcKeepTokenfromLPToken(lpBalance, pairData) {
+    const uniswapTotalSupply = new BN(
+      this.context.web3.utils.toWei(pairData.totalSupply.toString())
+    )
+    const keepLiquidityPool = new BN(
+      this.context.web3.utils.toWei(pairData.reserve0.toString())
```

Comment:
Should we assert that `reserve0` is really KEEP token, e.g. by checking `token0.symbol`?

```
if (pairData.token0.symbol !== "KEEP") {
    throw new Error(`unexpected token for reserve0: ${pairData.token0.symbol}`)
}
```

Reply:
I don't know, do you have a strong preference about it? We use a fixed lp pair address to fetch the data and KEEP token is already assigned to `reserve0`. I think that this `if` statement certainly wouldn't hurt to have but is it necessary?

Reply:
Nah, you're right. These are just two fixed pairs we know. Let's drop this idea.